data/transactions/verify/verifiedTxnCache.go

```diff
@@ -0,0 +1,292 @@
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	for i := range txgroup {
+		v.add(txgroup[i], groupCtxs[i])
+	}
+	return nil
+}
+
+// GetUnverifiedTranscationGroups compares the provided payset against the currently cached transactions and figure which transaction groups aren't fully cached.
+func (v *verifiedTransactionCache) GetUnverifiedTranscationGroups(txnGroups [][]transactions.SignedTxn, currSpecAddrs transactions.SpecialAddresses, currProto protocol.ConsensusVersion) (unverifiedGroups [][]transactions.SignedTxn) {
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	groupCtx := &GroupContext{
+		specAddrs:        currSpecAddrs,
+		consensusVersion: currProto,
+	}
+	unverifiedGroups = make([][]transactions.SignedTxn, 0, len(txnGroups))
+
+	for txnGroupIndex := 0; txnGroupIndex < len(txnGroups); txnGroupIndex++ {
+		signedTxnGroup := txnGroups[txnGroupIndex]
+		verifiedTxn := 0
+		groupCtx.minTealVersion = logic.ComputeMinTealVersion(signedTxnGroup)
+
+		baseBucket := v.base
+		for txnIdx := 0; txnIdx < len(signedTxnGroup); txnIdx++ {
+			txn := &signedTxnGroup[txnIdx]
+			id := txn.Txn.ID()
+			// check pinned first
+			entryGroup := v.pinned[id]
+			// if not found in the pinned map, try to find in the verified buckets:
+			if entryGroup == nil {
+				// try to look in the previously verified buckets.
+				// we use the (base + W) % W trick here so we can go backward and wrap around the zero.
+				for offsetBucketIdx := baseBucket + len(v.buckets); offsetBucketIdx > baseBucket; offsetBucketIdx-- {
+					bucketIdx := offsetBucketIdx % len(v.buckets)
+					if params, has := v.buckets[bucketIdx][id]; has {
+						entryGroup = params
+						baseBucket = bucketIdx
+						break
+					}
+				}
+			}
+
+			if entryGroup == nil {
+				break
+			}
+
+			if !entryGroup.Equal(groupCtx) {
+				break
+			}
+
+			if entryGroup.signedGroupTxns[txnIdx].Sig != txn.Sig || (!entryGroup.signedGroupTxns[txnIdx].Msig.Equal(txn.Msig)) || (!entryGroup.signedGroupTxns[txnIdx].Lsig.Equal(&txn.Lsig)) || (entryGroup.signedGroupTxns[txnIdx].AuthAddr != txn.AuthAddr) {
+				break
+			}
+			verifiedTxn++
+		}
+		if verifiedTxn != len(signedTxnGroup) || verifiedTxn == 0 {
+			unverifiedGroups = append(unverifiedGroups, signedTxnGroup)
+		}
+	}
+	return
+}
+
+// UpdatePinned replaces the pinned entries with the one provided in the pinnedTxns map. This is typically expected to be a subset of the
+// already-pinned transactions. If a transaction is not currently pinned, and it's can't be found in the cache, a errMissingPinnedEntry error would be generated.
+func (v *verifiedTransactionCache) UpdatePinned(pinnedTxns map[transactions.Txid]transactions.SignedTxn) (err error) {
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	pinned := make(map[transactions.Txid]*GroupContext, len(pinnedTxns))
+	for txID := range pinnedTxns {
+		if groupEntry, has := v.pinned[txID]; has {
+			pinned[txID] = groupEntry
+			continue
```

Comment:
what's the point of replacing an entry with itself?

Reply:
don't get confused between the current `v.pinned` and the newly built `pinned`.